python/trainers/buffer.py

```diff
@@ -0,0 +1,233 @@
+							raise BufferException("The batch size and training length requested for get_batch where" 
+								" too large given the current number of data points.")
+							return 
+						tmp_list = []
+						padding = np.array(self[-1]) * 0 
+						for end in range(len(self), len(self) % training_length , -training_length)[:batch_size]:
+							tmp_list += [np.array(self[end-training_length:end])]
+						if (leftover != 0) and (len(tmp_list) < batch_size):
+							tmp_list +=[np.array([padding]*(training_length - leftover)+self[:leftover])]
+						return np.array(tmp_list)
+					# TODO: decide if we need to reset the local buffer now ?	
+			def reset_field(self):
+				"""
+				Resets the AgentBufferField
+				"""
+				self[:] = []
+		def __str__(self):
+			return ", ".join(["'{0}' : {1}".format(k, str(self[k])) for k in self.keys()])
+		def reset_agent(self):
+			"""
+			Resets the AgentBuffer
+			"""
+			#TODO: There might be some garbage collection issues ?
+			for k in self.keys():
+				try:
+					self[k].reset_field()
+				except:
+					print(k)
+		def __getitem__(self, key):
+			if key not in self.keys():
+				self[key] = self.AgentBufferField()
+			return super(Buffer.AgentBuffer, self).__getitem__(key)
+		def check_length(self, key_list):
+			"""
+			Some methods will require that some fields have the same length.
+			check_length will return true if the fields in key_list 
+			have the same length.
+			:param key_list: The fields which length will be compared
+			"""
+			if len(key_list) < 2:
+				return True
+			l = None
+			for key in key_list:
+				if key not in self.keys():
+					return False
+				if ((l != None) and (l!=len(self[key]))):
+					return False
+				l = len(self[key])
+			return True
+		def shuffle(self, key_list = None):
+			"""
+			Shuffles the fields in key_list in a consistent way: The reordering will
+			be the same accross fields.
+			:param key_list: The fields that must be shuffled.
+			"""
+			if key_list is None:
+				key_list = list(self.keys())
+			if not self.check_length(key_list):
+				raise BufferException("Unable to shuffle if the fields are not of same length")
+				return
+			s = np.arange(len(self[key_list[0]]))
+			np.random.shuffle(s)
+			for key in key_list:
+				self[key][:] = [self[key][i] for i in s]
+				#memory leak here ?
+				# with l as [self[key][i] for i in s]:
+				# 	self[key].reset_field()
+				# 	self[key] += l
+				# self[key] = Buffer.AgentBuffer.AgentBufferField([self[key][i] for i in s])
+				# self[key].reorder(s)
+	def __init__(self):
+		self.global_buffer = self.AgentBuffer() 
+		super(Buffer, self).__init__()
+	def __str__(self):
+		return "global buffer :\n\t{0}\nlocal_buffers :\n{1}".format(str(self.global_buffer), 
+			'\n'.join(['\tagent {0} :{1}'.format(k, str(self[k])) for k in self.keys()]))
+	def __getitem__(self, key):
+		if key not in self.keys():
+			self[key] = self.AgentBuffer()
+		return super(Buffer, self).__getitem__(key)
+	def append_BrainInfo(self, info):
```

Comment:
stick to snake_case and not UpperCamelCase in function names

https://www.python.org/dev/peps/pep-0008/#function-names

blank lines

https://www.python.org/dev/peps/pep-0008/#blank-lines